R/dependencies.R

```diff
@@ -825,3 +825,85 @@ update_with_direct_deps <- function(project = ".",
   )
 }
 
+#' Recursively discover and build, check and install
+#' internal dependencies
+#'
+#' It discovers all dependencies starting from the repositories,
+#' determines the installation order and then
+#' builds, checks and installs them in order.
+#'
+#' @md
+#' @inheritParams rec_checkout_internal_deps
+#' @param artifact_dir (`character`) directory where build
+#'   tarball and logs go to
+#' @export
+#' @examples
+#' \dontrun{
+#' build_check_install_repos(
+#'   list(list(repo = "insightsengineering/teal", host = "https://github.com")),
+#'   feature = "main",
+#'   direction = "upstream",
+#'   local_repos = data.frame(repo = "insightsengineering/teal",
+#'   host = "https://github.com", directory = "../scratch1/teal/",
+#'   stringsAsFactors = FALSE)
+#' )
+#' build_check_install_repos(
+#'   list(list(repo = "maximilian_oliver.mordig/stageddeps.electricity",
+#'   host = "https://code.roche.com")),
+#'   feature = "master",
+#'   direction = "upstream",
+#'   #local_repos = data.frame(repo = "maximilian_oliver.mordig/stageddeps.electricity",
+#'   #host = "https://code.roche.com",
+#'   #directory = "../example_ecosystem/stageddeps.electricity/",
+#'   #stringsAsFactors = FALSE),
+#'   artifact_dir = "/tmp/test112"
+#' )
+#' }
+build_check_install_repos <- function(repos_to_process, feature,
+                                      direction = c("upstream", "downstream"),
+                                      local_repos = get_local_pkgs_from_config(),
+                                      verbose = 0,
+                                      artifact_dir = tempfile()) {
+  if (!dir.exists(artifact_dir)) {
+    dir.create(artifact_dir)
+  }
+
+  internal_deps <- rec_checkout_internal_deps(
+    repos_to_process, feature, direction, local_repos, verbose
+  )
+  deps <- get_true_deps_graph(internal_deps, direction = "upstream")
+  install_order <- get_install_order(deps[["upstream_deps"]])
+  install_order_paths <- vapply(
+    install_order, function(x) internal_deps[[hash_repo_and_host(x)]], character(1)
+  )
+
```

Comment:
This only builds, installs, tests upstream? I think we want this to do upstream and downstream dependent on direction? Looking at the issue:

"We would like to create a CI pipeline that installs and tests all packages in the staged dependencies and run a matrix test over different environments."

Reply:
No, but it uses the graph (where arrows point upstream) to compute the installation order. We can discover packages going upstream and/or downstream. However, to install, we only need the upstream graph.

Reply:
I updated the doc.